starknet_py/contract_abi_resolver.py

```diff
@@ -0,0 +1,131 @@
+# Needed because of string typed Contract
+# pyright: reportUndefinedVariable=false
+import warnings
+from enum import Enum
+from typing import Union, List, TypedDict, Tuple
+
+from starknet_py.net.client import Client
+from starknet_py.net.client_errors import ContractNotFoundError, ClientError
+from starknet_py.net.client_models import Abi, DeclaredContract
+from starknet_py.net.models import Address
+from starknet_py.proxy_check import (
+    ProxyResolutionError,
+    OpenZeppelinProxyCheck,
+    ArgentProxyCheck,
+    ProxyCheck,
+)
+
+
+class ProxyConfig(TypedDict, total=False):
+    """
+    Proxy resolving configuration
+    """
+
+    proxy_checks: List[ProxyCheck]
+    """
+    List of classes implementing :class:`starknet_py.proxy_check.ProxyCheck` ABC,
+    that will be used for checking if contract at the address is a proxy contract.
+    """
+
+
+def prepare_proxy_config(proxy_config: Union[bool, ProxyConfig]) -> ProxyConfig:
+    if isinstance(proxy_config, bool):
+        if not proxy_config:
+            return ProxyConfig()
+        proxy_config = ProxyConfig()
+
+    if "max_steps" in proxy_config:
+        warnings.warn(
+            "ProxyConfig.max_steps is deprecated. Contract.from_address always makes at most 1 step.",
+            category=DeprecationWarning,
+        )
+    proxy_checks = [OpenZeppelinProxyCheck(), ArgentProxyCheck()]
+    return {"proxy_checks": proxy_checks, **proxy_config}
+
+
+class ImplementationType(Enum):
+    """
+    Enum representing transaction types
+    """
+
+    CLASS_HASH = "class_hash"
+    ADDRESS = "address"
+
+
+class ContractAbiResolver:
+    """
+    Class for resolving abi of a contract
+    """
+
+    @staticmethod
+    async def resolve_direct(
+        address: Address,
+        client: Client,
+    ) -> Abi:
+        """
+        Returns abi of a contract directly from given address.
+        """
+        contract_class = await ContractAbiResolver._get_class_by_address(
+            address=address, client=client
+        )
+        return contract_class.abi or []
```

Comment:
What if we return `[]` as abi? Doesn't that make `Contract` instance created using it useless? Perhaps we shoud raise if we cannot get the abi or add some handling to usages of this method in `Contract`?

cc @Solpatium

Reply:
When is is it even possible to get None for abi? I think an empty list should always be returned from api 🤔

Reply:
I think it's a problem with 0.1.0 version of the rpc spec we currently support: It doesn't have abis at all. 
Even though, if the api returns empty list, it's still problematic in the same way as default `[]` is.

Reply:
Gateway does currently always return `abi`, but since it is not in RPC, `abi` is marked optional in `DeclaredContract`.
The statement `or []` in return was purely for typechecker not to throw errors.

Reply:
I'd do some more explicit handling, using default `[]` will cause issues down the road.